proposals/low-level-struct-improvements.md

```diff
@@ -383,11 +387,17 @@ The presence of `ref` fields means the rules around method arguments must match
 
 > For any method invocation `e.M(a1, a2, ... aN)`
 > 1. Calculate the *safe-to-escape* of the method return.
->     - Ignore the *ref-safe-to-escape* of arguments to `in / ref / out` parameters of `ref struct` types. The corresponding parameters *ref-safe-to-escape* are at most *return only* and hence cannot be returned via `ref` or `out` parameters.
+>     - Ignore the *ref-safe-to-escape* of arguments whose corresponding parameters have a *ref-safe-to-escape* of *return-only* or smaller.
+>     - Assume the method has a `ref struct` return type.
+> 2. All `ref` arguments of `ref struct` types must be assignable by a value with that *safe-to-escape*. This applies even when the `ref` argument matches a `scoped ref` parameter.
+
+> Further for any method invocation `e.M(a1, a2, ... aN)`
+> 1. Calculate the *safe-to-escape* of the method return.
```

Comment:
>*safe-to-escape* of the method return [](http://example.com/codeflow?start=19&length=37)

It's a little confusing that we're calling this the "*safe-to-escape* of the method return", and also in the previous set of bullets, but we ignore different arguments in each case.

I'm not sure how to make it clearer though. Perhaps just add a "where:" to the end of bullet 1 in each case.
> 1. Calculate the *safe-to-escape* of the method return *where*:

Reply:
I was really trying to lean on method invocation rules here and essentially re-use them. Cause really that is what this is doing, using the method invocation rules to calculate the narrowest possible return and validate it's okay to assign to every `ref` / `out` parameter. The method invocation rules are simple enough now, maybe just restate them here and note that this is effectively applyiyng method escape rules.

Reply:
@cston here is the version where I don't rely on re-using method invocation rules. Do you think this is clearer? 

> For any method invocation `e.M(a1, a2, ... aN)`
> 1. Calculate the minimum of the following:
>     - *calling method*
>     - The *safe-to-escape* of all arguments
>     - The *ref-safe-to-escape* of all ref arguments whose corresponding parameters have a *ref-safe-to-escape* of *calling method*
> 2. All `ref` arguments of `ref struct` types must be assignable by a value with that *safe-to-escape*.

Reply:
That sounds good. One concern though is step 2 refers to "a value with that *safe-to-escape*" but step 1 doesn't call the result a *safe-to-escape* value. Perhaps step 1 could be:
> Calculate a *safe-to-escape* from the minimum of:

Reply:
@RikkiGibson are you okay with this change of the MAMM language?

Reply:
I think of calling method as the "minimum" scope, so I found the language confusing.

Reply:
In general I'd like if we could avoid saying bigger, smaller, minimum, etc. for scope, and say narrower, wider, narrowest, widest. Otherwise mentally switching between the spec and implementation becomes very difficult.

I imagine that when we invest in this area of the compiler later on, we could consider introducing a scope type where "narrower" scopes are also considered "smaller" per the comparison operators. I think that's one of several aspects of the implementation which makes understanding it more difficult.

Reply:
Okay, updated this section.